replica/table.cc

```diff
@@ -814,6 +797,43 @@ future<> table::parallel_foreach_compaction_group(std::function<future<>(compact
     });
 }
 
+static std::pair<dht::token, dht::token> get_token_range(const dht::partition_range& pr) {
+    assert(!pr.is_singular());
+    auto start_token = pr.start() ? pr.start()->value().token() : dht::first_token();
+    auto end_token = pr.end() ? pr.end()->value().token() : dht::maximum_token();
+    return std::make_pair(start_token, end_token);
+}
+
+stop_iteration table::foreach_compaction_group_until(const dht::partition_range& pr, std::function<stop_iteration(compaction_group&)> action) {
+    if (pr.is_singular()) {
+        return action(compaction_group_for_token(pr.start()->value().token()));
+    }
+    auto [start_token, end_token] = get_token_range(pr);
+    for (auto token = start_token; token < end_token; ) {
+        auto& cg = compaction_group_for_token(token);
+        token = cg.next_compaction_group_token();
+        if (action(cg) == stop_iteration::yes) {
```

Comment:
we should reuse existing table::compaction_groups_for_token_range() but with modification to suit the foreach needs.

Reply:
why?  In fact I was thinking the other way around and replace the use of `compaction_groups_for_token_range` with foreach loops, like we did with sstable_set::all vs. sstable_set::for_each_sstable

Reply:
I was not actually interested in the interface, but rather correctness. compaction_group_for_token() is not to be used by reads.

Reply:
you can change compaction_groups_for_token_range() to have this for each style.